File: 3 Algorithms on Graphs/Week 4 Paths in Graphs 2/PA4/dijkstra/dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <limits>

using std::vector;
using std::queue;
using std::pair;
using std::priority_queue;
using std::numeric_limits;

struct Node {
  int index, distance;
  Node(int index = 0, int distance = 0): index(index), distance(distance) {}
};

struct cmp {
	bool operator()(Node first, Node second){
		return (first.distance > second.distance) - (first.distance < second.distance);
	}
};

int distance(vector<vector<int> > &adj, vector<vector<int> > &cost, int s, int t) {
  //write your code here
  vector<int> dist(adj.size(), numeric_limits<int>::max());
  dist[s] = 0;
  
  priority_queue<Node, vector<Node>, cmp> H;
  
  H.push(Node(s, dist[s]));
  while(!H.empty()) {
	Node u = H.top();
	H.pop();
	int u_index = u.index;

    for (int i = 0; i < adj[u_index].size(); i++) {
	  int v = adj[u_index][i];
	  if(dist[v] > dist[u_index] + cost[u_index][i]) {
	    dist[v] = dist[u_index] + cost[u_index][i];
        H.push(Node(v, dist[v])); 
	  }
	}	
  }
  
  if(dist[t] == numeric_limits<int>::max())
    return -1;
    
  return dist[t];
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  int s, t;
  std::cin >> s >> t;
  s--, t--;
  std::cout << distance(adj, cost, s, t);
}

File: 3 Algorithms on Graphs/Week 4 Paths in Graphs 2/PA4/dijkstra/dijkstra.py
#Uses python3

import sys
import queue

class Node(object):
    def __init__(self, index, distance):
        self.index = index
        self.distance = distance
        
    def __gt__(self, other):
        return self.distance > other.distance
    
    def __lt__(self, other):
        return self.distance < other.distance
    
    def __eq__(self, other):
        return self.distance == other.distance
        
def distance(adj, cost, s, t):
    #write your code here
    dist = [float('inf')] * len(adj)
    dist[s] = 0
    
    H = queue.PriorityQueue()
    H.put(Node(s, dist[s]))
    while not H.empty():
        u = H.get()
        u_index = u.index
        for v in adj[u_index]:
            v_index = adj[u_index].index(v)
            if dist[v] > dist[u_index] + cost[u_index][v_index]:
                dist[v] = dist[u_index] + cost[u_index][v_index]
                H.put(Node(v, dist[v]))
    
    if dist[t] == float('inf'):
        return -1
    
    return dist[t]


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    s, t = data[0] - 1, data[1] - 1
    print(distance(adj, cost, s, t))
